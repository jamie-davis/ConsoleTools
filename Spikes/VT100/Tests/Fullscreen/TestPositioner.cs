using System.Linq;
using TestConsoleLib.Testing;
using VT100.Attributes;
using VT100.ControlPropertyAnalysis;
using VT100.FullScreen;
using VT100.FullScreen.ControlBehaviour;
using VT100.FullScreen.ScreenLayers;
using VT100.Tests.Fakes;
using Xunit;

namespace VT100.Tests.Fullscreen
{
    public class TestPositioner
    {
        #region Types for test

        [Screen]
        [Border(BorderType.Normal)]
        [Background(VtColour.Blue)]
        [InputBackground(VtColour.Yellow)]
        class Layout : ILayout
        {
            #region Implementation of ILayout

            public event LayoutUpdated LayoutUpdated;

            #endregion

            [TextBox("Real Name")]
            [Border(BorderType.Normal)]
            [InputForeground(VtColour.Black)]
            public string Name { get; set; }

            [TextBox("Nickname")]
            [InputBackground(VtColour.Red)]
            public string NickName { get; set; }

            [Button("Default")]
            public void SetDefaults()
            {
                Name = "Default name";
                NickName = "Defaulty";
            }
            
            [Button("OK", ExitMode.ExitOnSuccess)]
            public bool OK()
            {
                return true;
            }
        }


        #endregion
        
        [Fact]
        public void ControlsArePositioned()
        {
            //Arrange
            var layout = new Layout() { Name = "Test", NickName = "Wolf" };
            var app = new FakeFullScreenApplication(layout, 50, 20);
    
            //Act
            app.Start();

            //Assert
            app.Console.GetDisplayReport().Verify();
        }
    }
}