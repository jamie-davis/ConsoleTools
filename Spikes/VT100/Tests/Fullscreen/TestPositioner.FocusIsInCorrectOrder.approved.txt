Initial Display                                     Focus Chain                                             
--------------------------------------------------- --------------------------------------------------------
|┌────────────────────────────────────────────────┐ Focus TextBoxControl "Nested text box" (col 21, row 6)  
|│                                                │ Focus TextBoxControl "Real Name" (col 14, row 11)       
|│ ┌────────────────────────────────────────────┐ │ Focus TextBoxControl "Nickname" (col 20, row 14)        
|│ │                                            │ │ Focus TextBoxControl "Favourite Colour" (col 20, row 16)
|│ │ Nested label     nested label value        │ │ Focus ButtonControl "Default" (col 13, row 18)          
|│ │                                            │ │ Focus ButtonControl "OK" (col 23, row 18)               
>│ │ Nested text box  nested value              │ │ Focus ButtonControl "Cancel" (col 28, row 18)           
|│ │                                            │ │                                                         
|│ └────────────────────────────────────────────┘ │                                                         
|│                                                │                                                         
|│ ┌────────────────────────────────────────────┐ │                                                         
|│ │Real Name  Test                             │ │                                                         
|│ └────────────────────────────────────────────┘ │                                                         
|│                                                │                                                         
|│ Nickname          Wolf                         │                                                         
|│                                                │                                                         
|│ Favourite Colour  Red                          │                                                         
|│                                                │                                                         
|│            ┏━━━━━━━┓ ┏━━┓ ┏━━━━━━┓             │                                                         
|│            ┃Default┃ ┃OK┃ ┃Cancel┃             │                                                         
|│            ┗━━━━━━━┛ ┗━━┛ ┗━━━━━━┛             │                                                         
|└────────────────────────────────────────────────┘                                                         
|                                                                                                           
|                                                                                                           
|                                                                                                           
 --------------------------------------------------                                                         
                      ^                                                                                     
Cursor position: Row 6, Column 21                                                                           
                                                                                               